internal/app/machined/pkg/controllers/k8s/static_pod_list.go

```diff
@@ -0,0 +1,177 @@
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"github.com/talos-systems/talos/pkg/machinery/resources/k8s"
+)
+
+// StaticPodListController renders all static pod definitions as a PodList and serves it as YAML via HTTP.
+type StaticPodListController struct {
+	staticPodListMu sync.Mutex
+	staticPodList   []byte
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *StaticPodListController) Name() string {
+	return "k8s.StaticPodListController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *StaticPodListController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: k8s.NamespaceName,
+			Type:      k8s.StaticPodType,
+			Kind:      controller.InputStrong,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *StaticPodListController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: k8s.StaticPodListType,
+			Kind: controller.OutputExclusive,
+		},
+		{
+			Type: k8s.StaticPodListURLType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+type pod map[string]interface{}
+
+type podList struct {
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+
+	Items []pod `json:"items" protobuf:"bytes,2,rep,name=items"`
+
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,3,opt,name=apiVersion"`
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *StaticPodListController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	go ctrl.serveStaticPodSpecs(ctx, r, logger)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		staticPods, err := r.List(ctx, resource.NewMetadata(k8s.NamespaceName, k8s.StaticPodType, "", resource.VersionUndefined))
+		if err != nil {
+			return fmt.Errorf("error listing static pods: %w", err)
+		}
+
+		pl := podList{
+			Kind:       "PodList",
+			APIVersion: "v1",
+		}
+
+		for _, staticPod := range staticPods.Items {
+			staticPodSpec := staticPod.(*k8s.StaticPod).TypedSpec()
+
+			pl.Items = append(pl.Items, staticPodSpec.Pod)
+		}
+
+		manifestContent, err := yaml.Marshal(pl)
+		if err != nil {
+			return fmt.Errorf("error rendering list of static pods as yaml: %w", err)
+		}
+
+		if err := r.Modify(ctx, k8s.NewStaticPodList(k8s.NamespaceName, k8s.StaticPodListResourceID), func(r resource.Resource) error {
+			r.(*k8s.StaticPodList).TypedSpec().Manifest = manifestContent
+
+			return nil
+		}); err != nil {
+			return fmt.Errorf("error modifying StaticPodList resource: %w", err)
+		}
+
+		ctrl.staticPodListMu.Lock()
+		ctrl.staticPodList = manifestContent
+		ctrl.staticPodListMu.Unlock()
+	}
+}
+
+func (ctrl *StaticPodListController) serveStaticPodSpecs(ctx context.Context, r controller.Runtime, logger *zap.Logger) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if err := ctrl.startServer(ctx, r, logger); err != nil {
```

Comment:
controllers are not allowed to touch `r controller.Runtime` from another goroutine, so this needs to be rewritten not to use goroutines. 

the HTTP server should be started in main goroutine, while `.Serve` might run in additional goroutine.

in general, probably server should start and stop with the controller